### ISSUE TITLE
Result.Failure should support ErrorType

### ISSUE THREAD

**Issue by jshier** [Alamofire/Alamofire#703]
Just started looking at the 2.0 API changes and the `Result` type really stuck out. It's a great encapsulation but I find it rather strange that it uses `NSError` rather than `ErrorType` to capture errors. Using `ErrorType` instead would allow much greater flexibility for developers by allowing them to create their own, customized error types. For instance, I'm currently developing against the CloudFlare API (api.cloudflare.com) and it uses a few HTTP response codes combined with details in the response JSON to indicate what error has occurred. I created a `CloudFlareError` enum that encapsulates all of the different errors. It'd be great to be able to encapsulate this error type in my generic serializer, rather than having to spit it out some other way. Was there a particular reason why only `NSError` was supported?

**Comment by cnoon** [comment]
The reason that `NSError` was originally used was not allow any Alamofire users to have to switch on the error types coming out of Alamofire. Since most of the errors are `NSError` objects generated by the URL Loading System, we decided to keep all errors generated by Alamofire logic to be `NSError` objects as well to make it easier for the consumer.

With that said, we did not spend enough time considering the power user that is creating their own response serializers. I think it is a great idea to wrap up Alamofire errors and then your own custom errors into your own custom enum to make it easier for yourself at the app level. For anyone that is willing to go to those lengths, the Result type was restricting you from doing that in a Swift-like manner. Because of this, I've updated the Result type to now take an ErrorType in the Failure case in e489c5fe.

Thanks for the awesome feedback! 🍻
